src/transformers/models/flava/modeling_flava.py

```diff
@@ -0,0 +1,2101 @@
+                multimodal_embeddings,
+                multimodal_output,
+            )
+
+        return FLAVAModelOutput(
+            image_embeddings=image_embeddings,
+            image_output=image_output,
+            text_embeddings=text_embeddings,
+            text_output=text_output,
+            multimodal_embeddings=multimodal_embeddings,
+            multimodal_output=multimodal_output,
+        )
+
+
+class FLAVACodebookBlock(nn.Module):
+    def __init__(self, in_size: int, out_size: int, num_layers: int, **kwargs):
+        super().__init__()
+
+        n_hid = out_size // 4
+        self.post_gain = 1 / (num_layers**2)
+
+        if in_size != out_size:
+            self.id_path = _build_codebook_conv2d(in_size, out_size, kernel_size=1)
+        else:
+            self.id_path = nn.Identity()
+
+        self.res_path = nn.Sequential(
+            OrderedDict(
+                [
+                    ("relu_1", nn.ReLU()),
+                    ("conv_1", _build_codebook_conv2d(in_size, n_hid, kernel_size=3)),
+                    ("relu_2", nn.ReLU()),
+                    ("conv_2", _build_codebook_conv2d(n_hid, n_hid, kernel_size=3)),
+                    ("relu_3", nn.ReLU()),
+                    ("conv_3", _build_codebook_conv2d(n_hid, n_hid, kernel_size=3)),
+                    ("relu_4", nn.ReLU()),
+                    ("conv_4", _build_codebook_conv2d(n_hid, out_size, kernel_size=1)),
+                ]
+            )
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.id_path(x) + self.post_gain * self.res_path(x)
+
+
+# Inspired by DALLE Encoder in https://github.com/openai/DALL-E/blob/5be4b236bc3ade6943662354117a0e83752cc322/dall_e/encoder.py#L42
+@add_start_docstrings(
+    """
+    The FLAVA's codebook model inspired from DALL-E's original encoder. Outputs raw hidden states and can be used to
+    generate image tokens for an image based on DALL-E's vocab. To be used to generate labels for MIM. Use
+    `get_codebook_indices` to get image tokens for an image.
+    """,
+    FLAVA_START_DOCSTRING.format(config="FLAVACodebookConfig"),
+)
+class FLAVACodebook(FLAVAPreTrainedModel):
+    base_model_prefix = ""
+    config_class = FLAVACodebookConfig
+    main_input_name = "pixel_values"
+    supports_gradient_checkpointing = False
+
+    def __init__(
+        self,
+        config: FLAVACodebookConfig,
+        **kwargs: Any,
+    ):
+        super().__init__(config)
+
+        self.config = config
+        self.num_groups = config.num_groups
+        self.input_channels = config.input_channels
+        self.num_blocks_per_group = config.num_blocks_per_group
+        self.hidden_size = config.hidden_size
+        self.vocab_size = config.vocab_size
+
+        num_layers = self.num_groups * self.num_blocks_per_group
+        output_conv = _build_codebook_conv2d(8 * self.hidden_size, self.vocab_size, kernel_size=1)
+
+        self.blocks = nn.Sequential(
```

Comment:
This part is completely unreadable. We don't use intermediate functions to build blocks usually, so you should just define each block separately. Then you can build the Sequential once all blocks are created.